File: Difficulty: Medium/Longest Sub-Array with Sum K/longest-sub-array-with-sum-k.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution{
    public:
    int lenOfLongSubarr(int A[],  int N, int K) 
    { 
        // Complete the function
        map<long long, int>sumMap;
        long long sum =0;
        int maxlen=0;
        for (int i=0; i<N;i++){
            sum+=A[i];
            if (sum ==K){
             maxlen=max(maxlen,i+1);   
            }
            long long rem = sum-K;
            if (sumMap.find(rem)!= sumMap.end()){
                int len = i-sumMap[rem];
                maxlen=max(maxlen,len);
            }
            if (sumMap.find(sum)==sumMap.end()){
                sumMap[sum]=i;
            }
            }
            return maxlen;
    } 

};

//{ Driver Code Starts.

int main() {
	//code
	
	int t;cin>>t;
	while(t--)
	{
	    int n, k;
	    cin>> n >> k;
	    int a[n];
	    
	    for(int i=0;i<n;i++)
	        cin>>a[i];
	   Solution ob;
	   cout << ob.lenOfLongSubarr(a, n , k)<< endl;
	    
	}
	
	return 0;
}
// } Driver Code Ends

File: Difficulty: Medium/Longest Sub-Array with Sum K/longest-sub-array-with-sum-k.java
//{ Driver Code Starts
// Initial Template for Java

/*package whatever //do not write package name here */

import java.io.*;
import java.util.*;

class Array {

    // Driver code
    public static void main(String[] args) throws IOException {
        // Taking input using buffered reader
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int testcases = Integer.parseInt(br.readLine());

        // looping through all testcases
        while (testcases-- > 0) {
            String line = br.readLine();
            String[] element = line.trim().split("\\s+");
            int sizeOfArray = Integer.parseInt(element[0]);
            int K = Integer.parseInt(element[1]);

            int arr[] = new int[sizeOfArray];

            line = br.readLine();
            String[] elements = line.trim().split("\\s+");
            for (int i = 0; i < sizeOfArray; i++) {
                arr[i] = Integer.parseInt(elements[i]);
            }

            Solution obj = new Solution();
            int res = obj.lenOfLongSubarr(arr, sizeOfArray, K);

            System.out.println(res);
        }
    }
}

// } Driver Code Ends


// User function Template for Java

class Solution {
    // Function for finding maximum and value pair
    public static int lenOfLongSubarr(int arr[], int n, int sum) {
        // Complete the function
        HashMap<Integer,Integer> h = new HashMap<>();
        int res = 0;
        int ps = 0;
        for(int i=0;i<n;i++){
            ps += arr[i];
            if(ps == sum){
                res = i+1;
            }
            if(!h.containsKey(ps)){
                h.put(ps,i);
            }
            if(h.containsKey(ps-sum)){
                res = Math.max(res,i-h.get(ps-sum));
            }
        }
        return res;
    }
}
